http/apache/parser.py

```diff
@@ -0,0 +1,167 @@
+    log_lines: List[str]
+
+
+def receive_with_prepended_length(sock: socket.socket) -> bytes:
+    response_bytes = sock.recv(4096)
+    # First line of response is its total size
+    expected_size_str, _, response_bytes = response_bytes.partition(b'\n')
+    expected_size = int(expected_size_str)
+    response_chunks = [response_bytes]
+    bytes_read = len(response_bytes)
+    while bytes_read < expected_size:
+        response_bytes = sock.recv(4096)
+        bytes_read += len(response_bytes)
+        response_chunks.append(response_bytes)
+    return b''.join(response_chunks)
+
+
+def nanny_version_matches(version: str) -> bool:
+    version_path = NANNY_RUN_DIR / 'version'
+    if version_path.is_file():
+        with open(version_path, 'r') as version_file:
+            running_nanny_version = version_file.read()
+        return running_nanny_version == version
+    return False
+
+
+def start_nanny(
+    port: int,
+    server_port_range_start: int,
+    server_instances: Optional[int],
+    version: str,
+) -> None:
+    p = subprocess.Popen(
+        [
+            './nanny.py', '--listen-port', str(port),
+            '--server-port-range-start', str(server_port_range_start),
+            '--stop-running-servers',
+            '--run-dir', '/var/run/apache-nanny/',
+            '--version', version,
+            *(
+                ['--max-instances', str(server_instances)]
+                if server_instances is not None else []
+            ),
+        ],
+        start_new_session=True,
+        stdin=subprocess.DEVNULL,
+        stdout=subprocess.DEVNULL,
+        stderr=subprocess.DEVNULL,
+    )
+    print(f'Running nanny process at pid {p.pid}')
+    time.sleep(1)
+
+
+def main():
+    argument_parser = argparse.ArgumentParser()
+    argument_parser.add_argument(
+        '--request-stream', required=True, help='File containing HTTP requests'
+    )
+    argument_parser.add_argument(
+        '--json-output', required=False, default=None,
+        help='File in which to write parsed requests as JSON'
+    )
+    argument_parser.add_argument(
+        '--load-balancer-port', required=True, type=int,
+        help='Port where load balancer around apache server is running'
+    )
+    argument_parser.add_argument(
+        '--load-balancer-version', required=True,
+        help='Version identifier for load balancer'
+    )
+    argument_parser.add_argument(
+        '--server-port-range-start', required=True, type=int,
+        help=(
+            'First port where server runs; further instances run on '
+            'subsequent ports'
+        )
+    )
+    argument_parser.add_argument(
+        '--server-instances', required=False, type=int, default=None,
+        help=(
+            'Number of server instances to run (default: use nanny CLI default)'
+        )
+    )
+    args = argument_parser.parse_args()
+
+    nanny_started = False
+    if not nanny_version_matches(args.load_balancer_version):
```

Comment:
Shouldn't this also kill the old nanny ? Perhaps that happens automatically

Reply:
I see it above now. I'd probably recommend adding a comment to that effect, just for clarity.

Alternatively, rename `start_nanny` to something that more fully describes what it does

Reply:
@caleblf